File: src/lib/api/connections.ts
import { invoke } from "./client";
import type { ConnectionProfile, SessionInfo } from "./types";

// Typed wrappers over the connection commands. The ONLY invoke site for this
// domain (DESIGN §9). Arg keys are camelCase; Tauri maps them to snake_case.
// `password` is memory-only — the profile has no password field by construction,
// so it rides as a transient arg on connect/test and is never persisted.
export const connectionsApi = {
  list: () => invoke<ConnectionProfile[]>("list_connections"),
  save: (profile: ConnectionProfile) => invoke<void>("save_connection", { profile }),
  remove: (id: string) => invoke<void>("delete_connection", { id }),
  test: (profile: ConnectionProfile, password?: string) =>
    invoke<void>("test_connection", { profile, password }),
  // `database` overrides the profile's — a Postgres database node opening its
  // own session, since a pg connection can never leave the database it opened.
  connect: (profileId: string, password?: string, database?: string) =>
    invoke<SessionInfo>("connect", { profileId, password, database }),
  disconnect: (sessionId: string) => invoke<void>("disconnect", { sessionId }),
};

File: src-tauri/src/commands/connections.rs
//! Thin connection command handlers: deserialize, call the service, map the
//! error. No logic here. Tauri maps the JS camelCase args to these snake_case
//! params (`profileId` → `profile_id`, `sessionId` → `session_id`).

use tauri::State;

use crate::config::connections::{self, ConnectionProfile};
use crate::drivers::types::SessionInfo;
use crate::services::connection_service;
use crate::state::AppState;
use crate::AppResult;

#[tauri::command]
pub fn list_connections(state: State<AppState>) -> AppResult<Vec<ConnectionProfile>> {
    connections::load_all(&state.paths)
}

#[tauri::command]
pub fn save_connection(profile: ConnectionProfile, state: State<AppState>) -> AppResult<()> {
    connections::save(&state.paths, &profile)
}

#[tauri::command]
pub fn delete_connection(id: String, state: State<AppState>) -> AppResult<()> {
    connections::delete(&state.paths, &id)
}

#[tauri::command]
pub async fn test_connection(
    profile: ConnectionProfile,
    password: Option<String>,
) -> AppResult<()> {
    connection_service::test_connection(&profile, password.as_deref()).await
}

#[tauri::command]
pub async fn connect(
    profile_id: String,
    password: Option<String>,
    state: State<'_, AppState>,
) -> AppResult<SessionInfo> {
    let profile = connections::load_one(&state.paths, &profile_id)?;
    connection_service::connect(&profile, password.as_deref(), &state.sessions).await
}

#[tauri::command]
pub async fn disconnect(session_id: String, state: State<'_, AppState>) -> AppResult<()> {
    connection_service::disconnect(&session_id, &state.sessions).await
}
